arch/arm/core/aarch32/cortex_a_r/thread_abort.c

```diff
@@ -0,0 +1,38 @@
+/*
+ * Copyright (c) 2020 Stephanos Ioannidis <root@stephanos.io>
+ * Copyright (c) 2016 Wind River Systems, Inc.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @file
+ * @brief ARM Cortex-A and Cortex-R k_thread_abort() routine
+ *
+ * The ARM Cortex-A and Cortex-R architectures provide their own
+ * k_thread_abort() to deal with different CPU modes when a thread aborts.
+ */
+
+#include <kernel.h>
+#include <kswap.h>
+
+extern void z_thread_single_abort(struct k_thread *thread);
+
+void z_impl_k_thread_abort(k_tid_t thread)
```

Comment:
I don't think Cortex A/R should enable CONFIG_ARCH_HAS_THREAD_ABORT. Can we delete this file and unset that Kconfig?

Please compare this to the base implementation
```
void z_impl_k_thread_abort(k_tid_t thread)
{
	__ASSERT((thread->base.user_options & K_ESSENTIAL) == 0U,
		 "essential thread aborted");

	z_thread_single_abort(thread);
	z_thread_monitor_exit(thread);

	if (thread == _current && !arch_is_in_isr()) {
		/* Direct use of swap: reschedule doesn't have a test
		 * for "is _current dead" and we don't want one for
		 * performance reasons.
		 */
		z_swap_unlocked();
	} else {
		z_reschedule_unlocked();
	}
}
```

The only difference is the missing:

```
	} else {
		z_reschedule_unlocked();
	}
```

Which AFAICT is incorrect.

Reply:
@andrewboie Thanks for reviewing.

My question is: is it really necessary to swap if the aborted thread is not the current thread?

In co-operative context switching case, unless `k_thread_abort` is meant to act like `k_yield`, it should not swap as long as the current thread is not aborted (though, does `z_thread_single_abort` unconditionally make the current thread unready? because, otherwise, the swap would be a no-op?); if preemptive context switching is enabled, the thread will be swapped in the system timer interrupt. Am I missing something here?

What I really need is "do not swap inside an interrupt/exception" -- in Cortex-A/-R (unlike Cortex-M which uses PendSV that effectively acts like deferred procedure call), the swap routine calls SVC, which immediately switches to SVC mode and exits through the common interrupt handler causing the exception stack to leak (exceptions use their own exception mode stack pointer).

If it is indeed necessary to swap even when the aborted thread is not the current thread, I could change the implementation as follows:

```
	if (!arch_is_in_isr()) {
		if (thread == _current) {
			/* Direct use of swap: reschedule doesn't have a test
			 * for "is _current dead" and we don't want one for
			 * performance reasons.
			 */
			z_swap_unlocked();
		} else {
			z_reschedule_unlocked();
		}
	}
```

Reply:
@andrewboie ping

Reply:
For now, I have updated the Cortex-A/-R-specific `k_thread_abort` implementation to align with the default kernel implementation, in terms of swap behaviour.